crates/starknet-devnet-core/src/starknet/mod.rs

```diff
@@ -1161,6 +1185,92 @@ impl Starknet {
             .expect("should get current UNIX timestamp")
             .as_secs()
     }
+
+    /// Impersonates account, allowing to send transactions on behalf of the account, without its
+    /// private key
+    ///
+    /// # Arguments
+    /// * `account` - Account to impersonate
+    pub fn impersonate_account(&mut self, account: ContractAddress) -> DevnetResult<(), Error> {
+        if self.config.fork_config.url.is_none() {
+            return Err(Error::UnsupportedAction {
+                msg: "Account impersonation is supported when forking mode is enabled.".to_string(),
+            });
+        }
+        self.cheats.impersonate_account(account);
+        Ok(())
+    }
+
+    /// Stops impersonating account.
+    /// After this call, the account, previously impersonated can't be used to send transactions
+    /// without its private key
+    ///
+    /// # Arguments
+    /// * `account` - Account to stop impersonating
+    pub fn stop_impersonating_account(&mut self, account: &ContractAddress) {
+        self.cheats.stop_impersonating_account(account);
+    }
+
+    /// Turn on/off auto impersonation of accounts that are not part of the state
+    ///
+    /// # Arguments
+    /// * `auto_impersonation` - If true, auto impersonate every account that is not part of the
+    ///   state, otherwise dont auto impersonate
+    pub fn auto_impersonate_account(&mut self, auto_impersonation: bool) {
+        self.cheats.auto_impersonate(auto_impersonation);
+    }
+
+    /// Returns true if the account is not part of the state and is impersonated
+    ///
+    /// # Arguments
+    /// * `account` - Account to check
+    fn is_account_impersonated(
+        state: &mut StarknetState,
+        cheats: &Cheats,
+        account: &ContractAddress,
+    ) -> DevnetResult<bool> {
+        let is_contract_already_in_state = state.is_contract_deployed_locally(*account)?;
+        if is_contract_already_in_state {
+            return Ok(false);
+        }
+
+        Ok(cheats.is_impersonated(account))
```

Comment:
if `is_contract_deployed_locally` returns `true` then the whole statement will return `true`, but have to be `false`.

Reply:
True, good that you checked. I guess that it could be:
```rust
Ok(!state.is_contract_deployed_locally(*account)? || cheats.is_impersonated(account))
```

But this leads me to the question: What if a user tries impersonating a locally deployed account?

Reply:
it will return on state.is_contract_deployed_locally

Reply:
But should that earlier be reported as an error?

Reply:
yes, will add a check

Reply:
IMO the short form of the condition is harder to understand.

Reply:
Completely missed this comment. I agree about the condition being harder to understand in the short form. Resolving.